### ISSUE TITLE
RET504 false-positive when nested functions reference the variable

### ISSUE THREAD

**Issue by Enegg** [astral-sh/ruff#14052]
I've searched for `RET504` before opening the issue.

```py
def outer() -> list[object]:
    @register
    async def inner() -> None:
        print(layout)

    layout = [...]
    return layout  # noqa: RET504
```
The rule reports the assignment to `layout` as unnecessary, and provides a fix by rewriting it as `return [...]`, which will break the `inner` function.

The rule doesn't trigger when `layout` is assigned to above the definition of the `inner` function, however in my use-case a decorator replaces the function with an object, which is then used to define `layout`, mandating the assignment to layout happens below the function definition.

---
```toml
[tool.ruff.lint]
preview = true
explicit-preview-rules = true
select = [ "ALL" ]
```

VSCode extension (`v2024.54.0`), `ruff==0.7.1`

**Comment by Enegg** [comment]
Sure. Here, note how the function `my_button` is replaced with the component passed to the decorator:
```py
from collections import abc
from dataclasses import dataclass, field
from typing import TypeAlias

# --------- library code ---------

@dataclass
class Component:
    custom_id: str
    disabled: bool = False


Layout: TypeAlias = abc.Sequence[abc.Sequence[Component]]


class Interaction:
    async def respond(self, content: str | None = None, *, components: Layout = ()) -> None: ...


Callback: TypeAlias = abc.Callable[[Interaction], abc.Awaitable[None]]


@dataclass
class Store:
    ids_to_callbacks: dict[str, Callback] = field(default_factory=dict)

    def register(self, component: Component, /) -> abc.Callable[[Callback], Component]:
        def wraps(callback: Callback) -> Component:
            self.ids_to_callbacks[component.custom_id] = callback
            return component
        return wraps

    def make_id(self) -> str: ...
    async def listen(self) -> None: ...

# --------- user code ---------

def create_layout(store: Store) -> Layout:
    @store.register(Component(custom_id=store.make_id()))
    async def my_button(inter: Interaction) -> None:
        my_button.disabled = True
        await inter.respond("Hello World", components=layout)

    layout = [[my_button]]
    return layout


async def command(inter: Interaction) -> None:
    store = Store()
    layout = create_layout(store)
    await inter.respond(components=layout)
    await store.listen()
 ```
And here are a few [examples](https://github.com/Enegg/discord-ui-store/blob/bf60916b10b5fe5cf0231e0de1c9746c65f7fc1d/examples/basic.py) from my library relying on this setup.

**Comment by joaoe** [comment]
The presented example is quite obscure to properly detect. Like, how can ruff know if the inner function goes out of scope or is kept globally ?

**Comment by Enegg** [comment]
Whether a reference to the inner function is stored somewhere is not relevant to the issue.

The rule seems unaware of the fact a variable can be referenced by a closure before it is defined.
